### PR TITLE
Predicates

### LINKED ISSUE #46: support ZKP-style predicate requests and selective disclosure
As raised in #18 

>          // TODO: how do you express zero-knowledge predicate requirements?

The presentation request should be able to request predicates (and specify the values against which credential values should be tested)

### PR DESCRIPTION
**Pull request by brentzundel**
This PR fixes #46 

Signed-off-by: Brent Zundel <brent.zundel@gmail.com>

### CODE REVIEW

**Comment by csuwildcat** [comment]
I'd like to have an extended discussion on this topic, because I was hoping we could use the JSON Schema filters for this purpose. I'd propose we dedicate a portion of the next call to this issue (cc @wyc)

**Comment by csuwildcat** [comment]
JSON Schema docs: https://json-schema.org/understanding-json-schema/

File: spec/spec.md
Comment (csuwildcat) [reply spec/spec.md:951]: If you're talking about checking a proof where the value itself is a Boolean, and you want to know the Boolean is true/false, wouldn't this also just be an enum or const?

**Comment by brentzundel** [comment]
> So after our call today, here were the main takeaways from my perspective:
> 
> * JSON Schema syntax is a superset of currently supported ZKP operations. We are going to recycle a subset of the syntax to apply to zero-knowledge protocol operations, due to many developers' familiarity with it. This means that instead of executing canonical albeit non-standardized JSON Schema validation rules, we will leave it to the ZKP implementers to figure it out, but we as spec writers will specify which subset of JSON Schema are appropriate to be interpreted as ZKP predicates (TO BE DONE cc @brentzundel). Furthermore, if ZKP circuit capabilities expand past JSON Schema's expressive capabilities, then we will need to revisit, and we accept this risk in the current version of the spec.

My first thought of a predicate operation that might not be expressable in JSON Schema turned out to be much closer to link secrets proofs than predicate proofs, so I think we're pretty good moving forward in this direction.

> * We still need to figure out the best way to support schemes such as [link secrets](https://forum.sovrin.org/t/proving-claim-ownership-without-sharing-did/860/2)

Agreed this still needs to be done, it is being tracked in issue #28

**Comment by csuwildcat** [comment]
@wyc has your requested change been fulfilled? Just waiting on merge until then.

**Comment by wyc** [comment]
Thanks @csuwildcat, I believe it has! Let me make the final approval

**Review by wyc — APPROVED** [review]
These changes look great, LGTM

**Pull request merged**
